File: THCS/2022_2023/QuangTri_2223/Cau3/SODB.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>

using namespace std;

// Hàm kiểm tra số nguyên tố
bool is_prime(int n) {
    if (n <= 1) return false; // Số nhỏ hơn hoặc bằng 1 không phải số nguyên tố
    if (n <= 3) return true;  // 2 và 3 là số nguyên tố
    if (n % 2 == 0 || n % 3 == 0) return false; // Loại trừ số chẵn và bội số của 3

    for (int i = 5; i * i <= n; i += 6) {
        if (n % i == 0 || n % (i + 2) == 0) return false;
    }
    return true;
}

// Hàm đếm số chữ số chẵn và lẻ
void count_even_odd_digits(int n, int &even_count, int &odd_count) {
    even_count = 0;
    odd_count = 0;

    while (n > 0) {
        int digit = n % 10; // Lấy chữ số cuối cùng
        if (digit % 2 == 0) {
            even_count++; // Chữ số chẵn
        } else {
            odd_count++; // Chữ số lẻ
        }
        n /= 10; // Bỏ chữ số cuối
    }
}

int main() {
    ifstream infile("SODB.INP");
    ofstream outfile("SODB.OUT");

    int N;
    infile >> N; // Đọc số lượng phần tử

    vector<int> A(N);
    for (int i = 0; i < N; ++i) {
        infile >> A[i]; // Đọc từng số nguyên
    }

    int special_count = 0;

    for (int i = 0; i < N; ++i) {
        int even_count = 0;
        int odd_count = 0;

        if (is_prime(A[i])) { // Kiểm tra nếu số nguyên là số nguyên tố
            count_even_odd_digits(A[i], even_count, odd_count); // Đếm chữ số chẵn và lẻ

            // Kiểm tra điều kiện số đặc biệt
            if (even_count != odd_count) {
                special_count++;
            }
        }
    }

    outfile << special_count << endl; // Ghi kết quả

    infile.close();
    outfile.close();

    return 0;
}

File: THCS/2022_2023/QuangTri_2223/Cau3/SODB.py
def is_prime(n):
    """Kiểm tra xem số nguyên n có phải là số nguyên tố hay không."""
    if n <= 1:
        return False
    if n <= 3:
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False
    
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    return True

def count_even_odd_digits(n):
    """Đếm số chữ số chẵn và lẻ trong số nguyên n."""
    even_count = 0
    odd_count = 0
    
    while n > 0:
        digit = n % 10
        if digit % 2 == 0:
            even_count += 1
        else:
            odd_count += 1
        n //= 10  # Chia nguyên cho 10 để loại bỏ chữ số cuối
    return even_count, odd_count

def main():
    with open("SODB.INP", "r") as infile:
        N = int(infile.readline().strip())  # Đọc số lượng phần tử
        A = list(map(int, infile.readline().strip().split()))  # Đọc danh sách các số nguyên

    special_count = 0

    for number in A:
        if is_prime(number):  # Kiểm tra số nguyên tố
            even_count, odd_count = count_even_odd_digits(number)  # Đếm chữ số chẵn và lẻ
            
            # Kiểm tra điều kiện số đặc biệt
            if even_count != odd_count:
                special_count += 1

    with open("SODB.OUT", "w") as outfile:
        outfile.write(str(special_count) + "\n")  # Ghi kết quả

if __name__ == "__main__":
    main()
